### PR TITLE
RSDK-3192 - Add robot reconnection logic

### PR DESCRIPTION
**Pull request by clintpurser**
This PR adds robot reconnection logic, I copied the logic from our python SDK, so it ought to be very similar to that.

There are a couple of fly by's included here, i will call them out below.

Some follow up work will be reworking our stream client to reconnect as well. this is under construction currently in [this](https://github.com/viamrobotics/viam-flutter-sdk/pull/53) - i will address in a follow up.

### CODE REVIEW

File: lib/src/resource/base.dart [hunk 1]
```diff
@@ -36,3 +39,9 @@ abstract class Resource {
     throw UnimplementedError();
   }
 }
+
+abstract class ResourceRPCClient<T extends Client> {
+  abstract ClientChannelBase channel;
+
+  T get client;
+}
```
Comment (clintpurser) [inline lib/src/resource/base.dart:47]: this abstract class helps us enforce that each component has a channel and client.

File: lib/src/components/movement_sensor/movement_sensor.dart [hunk 2]
```diff
@@ -52,54 +52,7 @@ abstract class MovementSensor extends Sensor {
...
-
-    try {
-      final orient = await orientation(extra: extra);
-      readings['orientation'] = orient;
-    } catch (exception) {
-      // TODO: Check if the exception is of a specific type and ignore or rethrow
-    }
-
-    return readings;
-  }
```
Comment (clintpurser) [inline lib/src/components/movement_sensor/movement_sensor.dart:102]: i _dont think_ this logic should have been here, moved to the client where i think it makes more sense with our current patterns.

File: test/unit_test/components/movement_sensor_test.dart [hunk 3]
```diff
@@ -267,6 +277,24 @@ void main() {
...
+          'linear_acceleration',
+          'compass',
+          'orientation'
+        ];
+        expect(result.keys, expectedKeys);
+
+        final expectedValues = [GeoPoint(latitude: 0.0), 0.0, Vector3(x: 0.0), Vector3(x: 0.0), Vector3(x: 0.0), 0.0, Orientation(oX: 0.0)];
+        expect(result.values, expectedValues);
+      });
+
```
Comment (clintpurser) [inline test/unit_test/components/movement_sensor_test.dart:297]: needed to add missing tests for `readings` now that its in the client.

File: lib/src/components/movement_sensor/movement_sensor.dart (hunk 2 above)
Comment (njooma) [reply lib/src/components/movement_sensor/movement_sensor.dart:102]: I think it should stay here, that way all subclasses of this resource will have this implementation out of the box (which I think is a good default)

**Review by njooma — CHANGES_REQUESTED** [review]

File: lib/src/robot/client.dart [hunk 4]
```diff
@@ -82,6 +102,63 @@ class RobotClient {
...
+          await _client.resourceNames(ResourceNamesRequest(), options: CallOptions(timeout: const Duration(seconds: 1)));
+          _connected = true;
+          break;
+        } catch (e) {
+          _connected = false;
+          await Future.delayed(const Duration(milliseconds: 100));
+        }
+      }
+      if (_connected) {
+        _logger.d('Robot still connected ${DateTime.now()}');
```
[the remark this answers is not in the document]
Comment (clintpurser) [reply lib/src/robot/client.dart:133]: that makes sense to me, I will remove this.

File: lib/src/components/movement_sensor/movement_sensor.dart (hunk 2 above)
Comment (clintpurser) [reply lib/src/components/movement_sensor/movement_sensor.dart]: SGTM I will move this back.

File: lib/src/robot/client.dart [hunk 5]
```diff
@@ -10,11 +13,19 @@ import '../resource/registry.dart';
...
 
+Logger _logger = Logger();
+
 /// The options that define the behavior of the [RobotClient].
 class RobotClientOptions {
   /// Options for connecting to the robot
   final DialOptions dialOptions;
 
+  /// The frequency (in seconds) at which to check if the robot is still connected. 0 (zero) signifies no connection checks
+  final checkConnectionInterval = 1;
```
Comment (cheukt) [inline lib/src/robot/client.dart:24]: I believe in other sdks the default is 10s

**Review by cheukt — APPROVED** [review]
LGTM, slightly out of scope here but can we add a simple `Connected` method so users can find if the robot is connected?

File: lib/src/robot/client.dart (hunk 5 above)
Comment (clintpurser) [reply lib/src/robot/client.dart:24]: ooh you are correct, i changed this to 1 for debugging purposes. Good catch, thank you!

**Comment by clintpurser** [comment]
@cheukt yeah I will add a `connected` method to the robot client... 👀

**Review by njooma — APPROVED** [review]

**Pull request merged**